src/jwt-decode-middleware.js

```diff
@@ -4,17 +4,34 @@ import authHeaderExtractor from './token-extractors/auth-header-extractor'
 import bodyExtractor from './token-extractors/body-extractor'
 import queryStringExtractor from './token-extractors/query-string-extractor'
 
+// Known errors in v8
+const JWT_ERRORS = [
+  'TokenExpiredError',
+  'NotBeforeError',
+  'JsonWebTokenError'
+]
+
 module.exports = function jwtDecodeMiddleware (options) {
   const tokenProperty = _.get(options, 'tokenProperty') || 'access_token'
-  options = _.omit(options, ['tokenProperty'])
+  const secret = _.get(options, 'secret')
   const bearerExtractor = authHeaderExtractor('bearer')
   const bodyTokenExtractor = bodyExtractor(tokenProperty)
   const queryTokenExtractor = queryStringExtractor(tokenProperty)
 
+  options = options || {}
+  options.ignoreExpiration = options.ignoreExpiration !== undefined ? options.ignoreExpiration : false
+  options.clockTolerance = options.clockTolerance !== undefined ? options.clockTolerance : 0
+  options = _.omit(options, ['secret', 'tokenProperty'])
+
   return (req, res, next) => {
-    const token = bearerExtractor(req) || bodyTokenExtractor(req) || queryTokenExtractor(req)
+    const token =
+      req.raw_jwt = bearerExtractor(req) || bodyTokenExtractor(req) || queryTokenExtractor(req)
 
-    req.jwt = token ? jwt.decode(token, options) : null
+    try {
+      req.jwt = token && jwt.verify(token, secret, options)
+    } catch (e) {
+      if (!JWT_ERRORS.includes(e.name)) throw e
```

Comment:
Why silence lib exceptions?

Reply:
The middleware is only supposed to decode, thus if the token is invalid it should not fail.

Reply:
That makes sense, but if ending up with an invalid (and silenced) token, is it then the applications concern to check explicitly for that? How would a user of this middleware be aware?

Reply:
Authorization middleware (on api-gateway) should validate the token. This middleware is used only to deserialize the `user` from the token. 

Initially, our assumption was that we should blindly trust the JWT, which I now think it's wrong. So, while we should enforce authorization here, we should only "trust" valid data.

Reply:
> I don't think it makes much sense to decode it in a middleware, and just silence errors.

The errors being thrown are just bad interface design from `jsonwebtoken`. If it would have had an explicit `verifySignature` and `verifyExpiry` method, the code would only have conditionals and no try / catch. 

> Aren't those errors something that should be blocking, unless explicitly unblocked?

They should be blocking in the authorization flow, which happens before a service is hit. Here we are just decoding (deserializing) data from the token, so while we still want to verify his authenticity, we don't want to block the request.

Reply:
It doesn't matter much really, but if you're choosing to silence those errors, maybe it makes sense to add it in the readme? This only means that we can't do anything about validation of these tokens on a service level.